File: LifeHtmlSite/Life.fs
module Life
open IntelliFactory.WebSharper
open IntelliFactory.WebSharper.Html5

[<JavaScript>]
type Cell = int * int
[<JavaScript>]
type Neighborhood = Set<Cell>
[<JavaScript>]
type Neighbors = Set<Cell>
[<JavaScript>]
type Generation = Set<Cell>

[<JavaScript>]
type IsAlive = bool
[<JavaScript>]
type LiveNeighborCount = int
[<JavaScript>]
type CellState = Cell * IsAlive * LiveNeighborCount

[<JavaScript>]
let neighborhood ((a,b):Cell) : Neighborhood = 
    [   for i in -1..1 do
        for j in -1..1 do
            yield a+i, b+j
    ] |> Set.ofList

[<JavaScript>]
let neighbors (cell:Cell) : Neighbors = neighborhood cell |> Set.remove cell

[<JavaScript>]
let cellState (liveCells:Generation) (cell:Cell) : CellState =
    let liveNeighborCount = neighbors cell |> Set.intersect liveCells |> Set.count
    let isCellAlive = liveCells |> Set.contains cell 
    (cell, isCellAlive, liveNeighborCount)

[<JavaScript>]
let survives (cellState:CellState) : bool = 
    match cellState with
        | _, _,    3 -> true
        | _, true, 2 -> true
        | _          -> false

[<JavaScript>]
let nextGeneration (livingCells:Generation) : Generation = 
    livingCells |> Set.map neighborhood 
                |> Set.unionMany 
                |> Set.map (cellState livingCells) 
                |> Set.filter survives 
                |> Set.map (fun (cell,_,_) -> cell)

File: PureJavascript/Life.js

var Life = (function() {
		var module = {};

		module.Cell = function (x, y) {
				this.x = x;
				this.y = y;
		};

		module.Cell.prototype.toKey = function() {
				return this.x + '_' + this.y;
		};

		module.CellState = function(cell, isCellAlive, liveNeighborCount) {
				this.cell = cell;
				this.isCellAlive = isCellAlive;
				this.liveNeighborCount = liveNeighborCount;
		};

		module.CellState.prototype.toKey = function() {
				return this.cell.toKey() + '_' + this.isCellAlive + '_' + this.liveNeighborCount;
		};

		module.nextGeneration = function(setOfLivingCells) {
				var cellsInPlay = new Set();
				setOfLivingCells.foreach( function(cell){
						var cellNeighborhood = neighborhood(cell);
						cellNeighborhood.foreach( function(neighboringCell){
								cellsInPlay.add(neighboringCell);
						});
				});

				var stateOfCellsInPlay = new Set();
				cellsInPlay.foreach( function(cellInPlay) {
						var stateOfCell = cellState(setOfLivingCells, cellInPlay);
						stateOfCellsInPlay.add(stateOfCell);
				});

				var survivingCells = new Set();
				stateOfCellsInPlay.foreach( function(state) {
						if (survives(state)) {
								var survivingCell = state.cell;
								survivingCells.add(survivingCell);
						}
				});

				return survivingCells;
		};

		function neighborhood (cell) {
				var x = cell.x;
				var y = cell.y;
				var neighborhood = new Set();
				for (var i=x-1; i<=x+1; i++){
						for (var j=y-1; j<=y+1; j++) {
								var neighbor = new module.Cell(i,j);
								neighborhood.add(neighbor);
						}
				}
				return neighborhood;
		}

		function neighbors(cell) {
				var myNeighbors = neighborhood(cell);
				myNeighbors.remove(cell);
				return myNeighbors;
		}

		function cellState(liveCells, cell) {
				var cellNeighbors = neighbors(cell);
				var liveNeighbors = cellNeighbors.intersection(liveCells);
				var liveNeighborCount = liveNeighbors.count();
				var isCellAlive = liveCells.contains(cell);
				return new module.CellState(cell, isCellAlive, liveNeighborCount);
		}

		function survives(state) {
				if (state.liveNeighborCount == 3) {
						return true;
				} else if (state.isCellAlive && state.liveNeighborCount == 2) {
						return true;
				} else {
						return false;
				}
		}

		return module;
})();